ReactAndroid/src/main/java/com/facebook/react/modules/settings/SettingsModule.java

```diff
@@ -0,0 +1,167 @@
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+package com.facebook.react.modules.settings;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+
+import com.facebook.react.bridge.Arguments;
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.facebook.react.bridge.ReactContextBaseJavaModule;
+import com.facebook.react.bridge.ReactMethod;
+import com.facebook.react.bridge.ReadableMap;
+import com.facebook.react.bridge.ReadableNativeMap;
+import com.facebook.react.bridge.WritableMap;
+import com.facebook.react.module.annotations.ReactModule;
+import com.facebook.react.modules.core.DeviceEventManagerModule;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+
+@ReactModule(name = SettingsModule.NAME)
+public class SettingsModule extends ReactContextBaseJavaModule  implements SharedPreferences.OnSharedPreferenceChangeListener{
+
+  @SuppressWarnings("WeakerAccess")
+  static final String NAME = "SettingsManager";
+  static final String DEFAULT_FILE_NAME = "ReactNative";
+
+  static private String sFilename = "";
+  private Boolean ignoringUpdates = false;
+  private SharedPreferences mPreferences;
+
+  public SettingsModule(ReactApplicationContext context) {
+    super(context);
+  }
+
+  @Override
+  public String getName() {
+    return SettingsModule.NAME;
+  }
+
+  private SharedPreferences getPreferences() {
+    if (mPreferences == null) {
+      if (SettingsModule.sFilename.length() > 0) {
+        mPreferences = getReactApplicationContext().getSharedPreferences(SettingsModule.sFilename, Context.MODE_PRIVATE);
+      } else if (getCurrentActivity() != null) {
+        mPreferences = getCurrentActivity().getPreferences(Context.MODE_PRIVATE);
+      } else {
+        mPreferences = getReactApplicationContext().getSharedPreferences(DEFAULT_FILE_NAME, Context.MODE_PRIVATE);
+      }
+      mPreferences.registerOnSharedPreferenceChangeListener(this);
+    }
+
+    return mPreferences;
+  }
+
+  static public void setFilename(String filename) {
+    sFilename = filename;
+  }
+
+  @Override
+  public void onCatalystInstanceDestroy() {
+    getPreferences().unregisterOnSharedPreferenceChangeListener(this);
+  }
+
+  @ReactMethod
+  public void setValues(ReadableMap map) {
+    ReadableNativeMap nativeMap = (ReadableNativeMap) map;
+
+    this.ignoringUpdates = true;
+    this.setValuesToPreference(nativeMap.toHashMap(), getPreferences());
+    this.ignoringUpdates = false;
+  }
+
+  @SuppressWarnings("unchecked")
+  public Map<String, Object> getConstants() {
+    Map<String, Object> map;
+
+    try {
+      map = (Map<String, Object>) getPreferences().getAll();
+
+      for (String key : map.keySet()) {
+        Object value = map.get(key);
+        if (value instanceof HashSet) {
+          ArrayList<Object> list = new ArrayList<>((HashSet<Object>) value);
+          map.put(key, list);
```

Comment:
this may cause trouble: the original map is being mutated - you don't want that. You will probably need an extra `Map` instance to cover the String Set case. Note you should be able to use `ArrayList<?> list = new ArrayList<>((Set<?>) value);` to avoid the unchecked casting complaint.